core/src/core/transaction.rs

```diff
@@ -1288,25 +1360,48 @@ impl From<Output> for OutputIdentifier {
 /// to produce a 32 byte message to sign.
 ///
 /// testnet4: msg = (fee || lock_height)
-/// mainnet:  msg = hash(fee || lock_height || features)
+/// mainnet:  msg = hash(features)                       for coinbase kernels
+///                 hash(features || fee)                for plain kernels
+///                 hash(features || fee || lock_height) for height locked kernels
 ///
 pub fn kernel_sig_msg(
 	fee: u64,
 	lock_height: u64,
 	features: KernelFeatures,
 ) -> Result<secp::Message, Error> {
+	if   features.is_coinbase()      && fee         != 0
+	 || !features.is_height_locked() && lock_height != 0 {
+		return Err(Error::InvalidKernelFeatures)
+	}
 	let msg = if global::is_testnet() {
 		let mut bytes = [0; 32];
 		BigEndian::write_u64(&mut bytes[16..24], fee);
 		BigEndian::write_u64(&mut bytes[24..], lock_height);
 		secp::Message::from_slice(&bytes)?
 	} else {
-		let hash = (fee, lock_height, features).hash();
+		let hash = if features.is_coinbase() {
+			(features).hash()
+		} else if features.is_plain() {
+			(features, fee).hash()
+		} else {
+			(features, fee, lock_height).hash()
```

Comment:
Any reason not to simply default the fee to 0 and lock_height to 0 and hash the tuple the same way every time?

```
kernel.fee() // just returns 0 if is_coinbase()
kernel.lock_height() // just returns 0 if !is_lock_height()
```

Reply:
Yes. The reason is it doesn't generalize nicely to more and more types.
Future types may have many more additional fields. With my approach , each type just hashes all relevant fields.
With the old approach, every type would have to encode defaults for all the possible fields of other types. That seems rather ugly and inefficient.

Reply:
I'm not convinced one big set of nested if/else statements are going to generalize nicely either though.
But :+1: we can always revisit this if/when it gets more complex.

Reply:
I'll rewrite the nested if/else as a match...
BTW, I tried to convert to enum using

/// Options for a kernel's structure or use
#[repr(u8)]
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub enum KernelFeatures {
        /// plain kernel has fee, but no lock_height
        PLAIN         = 0,
        /// coinbase kernel has neither fee nor lock_height (both zero)
        COINBASE      = 1,
        /// absolute height locked kernel; has fee and lock_height
        HEIGHT_LOCKED = 2,
}

...

impl KernelFeatures {
        /// convert from u8
        pub fn from_u8(num: u8) -> Option<KernelFeatures> {
                match num {
                        0 => Some(KernelFeatures::PLAIN),
                        1 => Some(KernelFeatures::COINBASE),
                        2 => Some(KernelFeatures::HEIGHT_LOCKED),
                        _ => None
                }
        }
        /// convert to u8
        pub fn to_u8(self) -> u8 {
                match self {
                        KernelFeatures::PLAIN => 0,
                        KernelFeatures::COINBASE=> 1,
                        KernelFeatures::HEIGHT_LOCKED => 2,
                }
        }

but I got huge amount of borrowing errors and gave up.
there must be a better way to use enums...

Reply:
any borrow errors you might have seen here here are just because your `to_u8` function is trying to capture the instance. Just use a reference to self instead:

```
pub fn to_u8(self) -> u8
```
to
```
pub fn to_u8(&self) -> u8
```

Reply:
I tried with https://raw.githubusercontent.com/tromp/tromp.github.io/master/transaction.rs
but still get tons and tons of errors, such as

error[E0507]: cannot move out of borrowed content                                                                          
   --> core/src/core/transaction.rs:244:3                                                                                  
    |                                                                                                                      
244 |         self.features.is_coinbase()                                                                                  
    |         ^^^^^^^^^^^^^ cannot move out of borrowed content                                                            
                                                                                                                           
error[E0507]: cannot move out of `self.features` which is behind a `&` reference                                           
   --> core/src/core/transaction.rs:244:3                                                                                  
    |                                                                                                                      
243 |     pub fn is_coinbase(&self) -> bool {                                                                              
    |                        ----- help: consider changing this to be a mutable reference: `&mut self`                     
244 |         self.features.is_coinbase()                                                                                  
    |         ^^^^^^^^^^^^^                                                                                                
    |         |                                                                                                            
    |         cannot move out of `self.features` which is behind a `&` reference                                           
    |         `self` is a `&` reference, so the data it refers to cannot be moved                                          
...
error[E0382]: use of moved value: `features`                                                                               
    --> core/src/core/transaction.rs:1396:44                                                                               
     |                                                                                                                     
1396 |     if features.is_coinbase() && fee != 0 || !features.is_height_locked() && lock_height != 0 {                     
     |        -------- value moved here              ^^^^^^^^ value used here after move                                   
     |                                                                                                                     
     = note: move occurs because `features` has type `core::transaction::KernelFeatures`, which does not implement the `Copy` trait

Reply:
all of your functions in kernelfeatures `is_coinbase, is_plain, is_height_locked` need to be `&self` as well

Reply:
When I make them &self, then I need to change
        pub fn is_coinbase(self) -> bool {
                self == KernelFeatures::COINBASE
to 
        pub fn is_coinbase(&self) -> bool {
                *self == KernelFeatures::COINBASE

And then I get errors like

error[E0507]: cannot move out of borrowed content                                                                          
   --> core/src/core/transaction.rs:265:56                                                                                 
    |                                                                                                                      
265 |         let msg = kernel_sig_msg(self.fee, self.lock_height, self.features)?;                                        
    |                                                              ^^^^^^^^^^^^^ cannot move out of borrowed content       
                                                                                                                           
error[E0507]: cannot move out of `self.features` which is behind a `&` reference                                           
   --> core/src/core/transaction.rs:265:56                                                                                 
    |                                                                                                                      
264 |     pub fn msg_to_sign(&self) -> Result<secp::Message, Error> {                                                      
    |                        ----- help: consider changing this to be a mutable reference: `&mut self`                     
265 |         let msg = kernel_sig_msg(self.fee, self.lock_height, self.features)?;                                        
    |                                                              ^^^^^^^^^^^^^                                           
    |                                                              |                                                       
    |                                                              cannot move out of `self.features` which is behind a `&` reference
    |                                                              `self` is a `&` reference, so the data it refers to cannot be moved